### DISCUSSION TITLE
Built-in resolution changing

### DISCUSSION THREAD

**Discussion by Ariane-B** [LizardByte/community#150] [Sunshine] [answered]
## The problem is a bit hard to solve manually

I know [GameStream Launchpad](https://github.com/cgarst/gamestream_launchpad) and [QRes](https://www.majorgeeks.com/files/details/qres.html) exist, but while I used to use GameStream Launchpad with Playnite back when I used Nvidia's proprietary GameStream, and it worked fine, I ended up giving up on it with Sunshine.

I saw a few support threads, saying to use the .bat files as detached commands, or to put the .bat file's command itself into Sunshine and forgoing the .bat file altogether, but I haven't been able to get it to work at all. Granted, I'm pretty bad at this kind of stuff AND I didn't put hours and hours into it, but yeah. It's not been easy.

In the mean time, I've been manually switching my monitor between 4K for games and 1440p for work (because 4K on a 1440p monitor makes some text really ugly). It works, but it's a bit of a pain in the butt to sit down to play, notice the graphics are kinda off, realize I forgot to switch the resolution, have to close the game and Sunshine, get up, change the resolution, and start over.

## My suggestion

It seems like the kind of thing that people would want to do often: stream from a PC whose monitor 1080p/1440p/ultrawide or whatever to their 4K TV.

So I think it'd be a really neat feature if we had a toggle in each app to automatically switch the host PC to the resolution the client asked to stream at, and then switch back once the connection is terminated.

I'd make the previously mentionned "Match requested resolution on host PC during stream" the default, but I don't see a reason why it couldn't be customized with specific resolutions. For instance, you could set up an app to open Playnite at 1080p and another one for 4K. You could choose on the client depending on how demanding the game you want to play is or whether you're currently streaming locally or over the Internet.

**Comment by KuleRucket** [comment]
I guess you could integrate QRes functionality, but then you have to do it differently on Linux and integrate xrandr for x11 or differently in Wayland depending on the compositors. E.g wlr-randr, gnome-randr etc. Then for Mac I have no idea.

It's much easier to just provide good instructions on running third-party tools via detached commands to accomplish the same thing.

**Comment by Ariane-B** [comment]
The problem is that then you have no control over the user friendliness of the tools. Guides can only help people to a degree when they're dealing with complex things. 

Sure, I'm only a front-end Web developer. But I'm still a developer. And yet that stuff is currently mostly over my head.

Clear instructions in the docs could help. They could certainly get things to a level where I would be comfortable giving it another shot.

But pathetic as I might be, I still very much think I'm way above the average gamer in terms of being comfortable with CLI stuff. So I'm not sure it's sufficient in the long term if Sunshine has ambitions of appealing to the general public eventually. 

Maybe the answer could be some sort of plugin structure. It could evolve further in the future, but the "plugins" could start as simply as a curated list of commands you can easily copy paste and add to your apps, with clear explanations of any parameters?

Then maybe a step 2 in complexity could be a few spots in the interface where a plugin author could inject HTML/CSS/JS, plus a few hooks/events in the program itself where they could run commands, synchronously or asynchrously. Kinda WordPress like, I guess. 

With the resolution switching plugin, you'd be prompted to install one of the required resolution switching programs, kinda like peer dependencies in npm, and it'd give you an intuitive interface for it and a good, working default config.

And then I guess step 3 would be a real full fledged framework/API/etc?

That's just one idea though. Maybe there's something much better to do to solve that issue. 

Tbh, resolution switching arises from such a common use case it's kinda bewildering to me that it wasn't included with Nvidia's proprietary streaming stuff to begin with. They could've even integrated it with their driver-side DLSS-like upscaler for more wow factor. "4K gaming even on the 3050!" Or something. 

I don't know what exactly, but I hope there's a way to integrate it in a way that's as user friendly as possible.

**Reply by ReenigneArcher** [reply]
I'm not saying there won't ever be an integration like this, but integrating too many features greatly increases the maintenance burden on the project. If other tools exist to accomplish the need, it's probably less likely to get integrated. An example of this is using ViGEmBus for controller support on Windows.

**Reply by Ariane-B** [reply]
That's fair. Maintenance is paintenance sometimes. I guess I'm just a little frustrated because of how rudimentary the existing tools are.

**Comment by KuleRucket** [comment]
In case it helps, this is what I have in Linux/X11 to switch resolution, scale and refresh rate for my TV on connect and then reset it ready for desktop use on exit. I guess QRes would be similar.
Do:
`xrandr --output DP-0 --mode 2560x1440 --refresh 120.00 --scale 0.75x0.75`
Undo:
`xrandr --output DP-0 --mode 2560x1440 --refresh 144.00 --scale 1x1`

I include 1440p both times because I have another profile that does 4k/60 and occasionally it doesn't switch back on exit

**Reply by Ariane-B** [reply]
That probably will help. I'm kind of out of patience for tinkering today, but I'll make sure to give QRes another go soon-ish. :)

**Reply by Ariane-B** [reply]
Tried QRes. Seems that my desktop monitor can't do 4K 60 Hz, but by adding a custom resolution to the Nvidia control panel, it works. It seems my monitor treats anything unsupported as its native resolution, or something.

Anyway it seems to work now! Never managed to add QRes to my PATH, but by putting the whole path to the executable, it changes the resolution fine!

**Comment by khoanguyen-3fc** [comment]
But doesn't it only change Desktop resolution, exclusive fullscreen Games would have a separated resolution?

In my setup, host resolution always 1080 (both in-game and desktop), and when I'm on slow bandwidth internet, I would downscale stream to 720 (config on Moonlight), and it doesn't look really bad.

**Reply by Ariane-B** [reply]
That works downwards, but not upwards. Games won't allow you to run at 4K when the system resolution is 1440P. And even if I run the stream at 4K, Moonlight can't create pixels out of nothing. 

They might have a render scale setting, but usually that'll be percent-based and it will go in stops. You can go 150% bigger or 200% bigger but you can't do exactly 4K. 

And even if you did that, it wouldn't help much. You'd probably still receive a 1440P signal, and no matter what, the interface would still be rendered at 1440P.

**Reply by ReenigneArcher** [reply]
> Games won't allow you to run at 4K when the system resolution is 1440P.

I 'm not sure if that's true. https://nvidia.custhelp.com/app/answers/detail/a_id/5280

It's been a while but I streamed Ancestors in 4k while i only have 1080p monitors.

**Reply by Ariane-B** [reply]
Hmm that's interesting. I'll have to take a look the next time I play

**Reply by khoanguyen-3fc** [reply]
I'm sure that I can set in-game resolution to 4k in exclusive fullscreen mode, when desktop in 1080, with a 4k dummy HDMI.

**Reply by Ariane-B** [reply]
Tried the Nvidia control panel thing. It worked, giving me the option to set a game to 4K, for a while, even though games always started at 1440p and I had to set them to 4K manually on every launch, and then started producing really ugly aliased images, and the 4K option was available but it went back to 1440p whenever I selected it. Not sure what happened.

**Reply by Ariane-B** [reply]
That would be amazing but you should create a new thread.

**Reply by ReenigneArcher** [reply]
> That would be amazing but you should create a new thread.

Why? It's close enough to this feature request to stay here.

**Reply by Ariane-B** [reply]
Well if you say so I certainly won't contradict you!

**Reply by Ariane-B** [reply]
Oh yeah it's an amazing idea. And it could basically act as a first iteration of the user-friendly option I dream of.

**Reply by AkazaRenn** [reply]
I would like to see that happening, too! I have a bunch of devices all of which offer different resolutions and even different aspect ratios, if we can use the information to adjust the host's resolution according it would truly be amazing.

**Comment by fmiceli24** [comment]
I have set up this with some of the recommended tools and playnite. The problem I have is how to automatically recognize the resolution requested by the moonlight client in order to pass it as a parameter to my do/undo scripts. If I can access this then I can share my strategy to automatic resolution changing.

**Comment by ReenigneArcher** [comment]
There is a script developed by one of our contributors to automate resolution changing on Windows, listed in the nightly docs.

https://docs.lizardbyte.dev/projects/sunshine/en/nightly/about/app_examples.html

**Reply by fmiceli24** [reply]
Yes. There is no issue with that. But what I want to do is change the resolution based on what the Moonlight device asks.
Where can I find the resolution Moonlight is asking so I can adjust this dynamically on my resolution changing script/task?

**Reply by ReenigneArcher** [reply]
Re-read the link I sent. There's a script that does exactly what you're asking for.

**Reply by fmiceli24** [reply]
Thanks @ReenigneArcher! Sorry my dislexia kicked in and I missed that link!

Here is the link BTW to save others a few clicks: [Nonary/ResolutionAutomation](https://github.com/Nonary/ResolutionAutomation/)

**Reply by Skyro620** [reply]
I can't get the resolutionautomation working. I'm assuming it reads the "virtual desktop" resolution to switch to it but my sunshine.log shows my virtual desktop as the same resolution as my host resolution.

**Comment by JohnnySixStrings** [comment]
I'm want to write a windows background service to handle the resolution changing for windows with an installer. Is there any better way than pulling from the logs to find out what resolution and refresh rate the client requested? Also is there any way to tell what process it is opening? I think could do this in C# and might even be able to pull it off on Linux as well.  No promises.

**Reply by ReenigneArcher** [reply]
This kind of already exists. [Nonary/ResolutionAutomation](https://github.com/Nonary/ResolutionAutomation/)

Also, it'll be a bit redundant once we release the next version which can pass arguments to the global prep commands from the client... but I guess if you wanted to write a wrapper for qres and/or whatever tools on Linux you could and instruct people to use it as a global prep command... But if you wanted to run it as a service, then I don't see any better way than using the logs. There's nothing in the API for this right now.

**Reply by Ariane-B** [reply]
@ReenigneArcher that looks amazing! Are you considering an official, "easy" plugin system in the future?

If that thing could be enabled from a slick, idiot-proof interface, I could see myself recommending Sunshine to pretty much anyone!

**Reply by ReenigneArcher** [reply]
We have discussed a plugin system a few times, but no work has been done on it.  It's not such a trivial thing to add at this point.

**Reply by Ariane-B** [reply]
Oh yeah I bet. A future plan of it already makes me happy enough. :)

**Reply by ns6089** [reply]
Plugin system in an opensource project doesn't make much sense anyway, Unless we want to support proprietary closed sourced plugins, which I don't think we do.

**Reply by TheElixZammuto** [reply]
Having a repository of the most common start/stop/detatched commands is not a bad idea tho. While not a full plugin system, it might work for a large chunk of usescases

**Reply by ReenigneArcher** [reply]
@ns6089 The big benefit would be that the maintenance burden falls back on the original contributor of that plugin, instead of increasing the burden on the few active developers... especially for a niche feature that not that many people use. (not saying resolution changing isn't a popular request though, because it definitely is)

**Reply by JohnnySixStrings** [reply]
@ReenigneArcher  Thanks for the reply. I would rather not do redundant work if this already in the works with passing arguments to the global prep commands.  Yes I have tried the Resolution Automation and the experience left me wanting. I looked through the code for it to decide if I could do something similar in c# with a windows service. Which I could later port to Linux.  I might still end up writing just cause I had fun last night porting nonary's work to c#.

**Reply by ReenigneArcher** [reply]
@JohnnySixStrings you can pass variables to prep commands now (in nightly build)

**Reply by Stono** [reply]
@ReenigneArcher where's the PR/Doco for that as i'm interested.  Does it pass dynamic variables, like the request resolution for example?

**Reply by ReenigneArcher** [reply]
@Stono https://docs.lizardbyte.dev/projects/sunshine/en/latest/about/guides/app_examples.html

**Reply by Stono** [reply]
Perfect, thanks a lot

**Reply by lsalazarm99** [reply]
The link is broken. This is the one working: https://docs.lizardbyte.dev/projects/sunshine/en/latest/about/guides/app_examples.html.

**Comment by psiberfunk** [comment]
It would be good if the documentation recommended something other than abandonware as a solution.  Something written for the Win95 era doesn't strike me as a great example.. could break any time.  Some slightly more modern alternatives if Sunshine doesn't want to build this in  (which still feels like the right thing to do... )  :

-https://github.com/gurnec/SetScreenResolution
-http://www.nirsoft.net/utils/nircmd.html
-https://www.entechtaiwan.com/util/multires.shtm

**Reply by ReenigneArcher** [reply]
All of these are also abandonware

https://github.com/gurnec/SetScreenResolution... last updated in 2016
http://www.nirsoft.net/utils/nircmd.html... last updated in 2019
https://www.entechtaiwan.com/util/multires.shtm... last updated in 2006

Fact is, these APIs don't change that often. Use whatever tool you want. QRES is just an exmaple, no one is forcing you to use it.

**Comment by psiberfunk** [comment]
P.S.  if you want to ALSO change HDR state.. you can run `nircmd.exe sendkeypress "lwin+alt+b"`  , which makes this utility a really easy win, as you can set resolution and toggle HDR state.. (although it would be ideal to set rather than toggle)...   Annoyingly though, despite me trying to figure out why.. sometimes it doesn't like the 2 commands in rapid succession, and this isn't really reliable to do both switches.


**Update:**

In case anyone finds this.. i used a combination of multires + nircmd to accomplish the task of toggling HDR and 1440/4k resolution like this:

![image](https://github.com/LizardByte/Sunshine/assets/19777571/d4f03896-906a-473f-b6cb-9e7b52f9fd15)

For some reason the nircmd resolution switch function worked fine by itself (e.g. `cmd /C nircmd.exe setdisplay 2560 1440 32`) but when chained with the keypress command refused to work properly.  I spent an hour trying to figure out why and eventually gave up and used 2 different add-on commands (MultiRes + NirCmd) as I show above.  Still wish I didn't have to though.. wasted 2 hours in frustration.

It would be ideal if Sunshine had a "switch resolution" and "switch HDR state" on app launch...  :).   Built-in functionality would avoid all this struggle to find the magic combinations.

**Comment by psiberfunk** [comment]
> It would be good if the documentation recommended something other than abandonware as a solution. Something written for the Win95 era doesn't strike me as a great example.. could break any time. Some slightly more modern alternatives if Sunshine doesn't want to build this in (which still feels like the right thing to do... ) :
> 
> -https://github.com/gurnec/SetScreenResolution -http://www.nirsoft.net/utils/nircmd.html -https://www.entechtaiwan.com/util/multires.shtm

2019 is at least relatively recent <shrug>.  But anyways, my point is that it's hard to find a single tool to do the HDR + resolution changing, and when you try to chain commands together you get unreliable results based on, as far as I can tell, race conditions... where if the second command executes too quickly after the first, it doesn't work properly.. i'm not really quite sure how to fix that, since the command executes and completes immediately, but it does take the graphics driver and monitor a few seconds to re-handshake at a new resolution/setting.

  Since fundamentally sunshine already has lots of resolution-related settings associated with encoding, it would seem to be a natural extension to have native resolution/HDR API calls as part of the app launching to make this experience less error-prone and guess-and-check.   

Is it something that would be considered for implementation ?

**Comment by vvuk** [comment]
I'm actually using this nirsoft tool -- [MultiMonitorTool](https://www.nirsoft.net/utils/multi_monitor_tool.html). It has the nice ability to save full configs, so I used the tool to save a "Default" config, and then my Do/Undo commands are:

```
Do: cmd /C C:\path\to\MultiMonitorTool.exe /SetMonitors "Name=GBT3204 Width=%SUNSHINE_CLIENT_WIDTH% Height=%SUNSHINE_CLIENT_HEIGHT% DisplayFrequency=%SUNSHINE_CLIENT_FPS%"
Undo: cmd /C C:\path\to\MultiMonitorTool.exe /LoadConfig C:\path\to\SavedDefaultConfig.cfg
```

seems to work well.  To go a little fancier I'd probably make the "do" step save the current config for undo to load, instead of going back to the primary one and put the whole thing into a single .bat file for easier maintenance. I'd also probably use the tool to turn off my second monitor while streaming. The nice thing about this tool is that it can make a full config change in one go, so reduces race conditions etc problems.

Parsec has "virtual displays" that it sets up just for streaming, so that your can have your actual monitor blanked. Would be nice to investigate something like this as well.

However, I do agree with this:
> Since fundamentally sunshine already has lots of resolution-related settings associated with encoding, it would seem to be a natural extension to have native resolution/HDR API calls as part of the app launching to make this experience less error-prone and guess-and-check.

A modern implementation for Windows using the CCD API wouldn't be too difficult. I believe it has to be CCD though in order to be able to set things like HDR color.

**Comment by httpjamesm** [comment]
Is there any movement on this? I have a mixture of 16:10 and 16:9 devices so the ability to automatically change resolution based on the client requested resolution would be fantastic.

**Comment by snipem** [comment]
Interesting that this is still unsolved for Moonlight, GeForce Experience and Sunshine. Sticking to Steam Link for now.

**Comment by ReenigneArcher** [comment, marked as answer]
This is integrated in Sunshine (for Windows) since https://github.com/LizardByte/Sunshine/pull/3441

To add support for Linux or macOS, please submit contributions to https://github.com/LizardByte/libdisplaydevice

**Reply by brybalicious** [reply]
I am using this now. It seems to work, selecting the right custom display resolution. Is there a way to have it switch back to the prior resolution (1440p for my monitor) once the stream is ended?

Moreover, is there any way to have Sunshine automatically turn off the primary monitor while streaming e.g. to the 4k TV?
